Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+      producing the appropriate IR.
+    * Add support for extracting the fields for the returned struct value of a call (in registers or on
+      stack), producing the appropriate IR.
+    * Add support for assembling non-matching fields into registers for call args and returns.
+    * For arm64, add support for loading non-promoted or non-local structs with ldp
+    * See [Extracting and Assembling Structs](#Extract-Assemble).
+    * The removal of each of these pessimizations should result in improved code generation
+      in cases where previously disabled optimizations are now enabled.
+
+### Fully Enable Struct Optimizations
+
+Most of the existing places in the code where structs are handled conservatively are marked
+with `TODO-1stClassStructs`. This work item involves investigating these and making the
+necessary improvements (or determining that they are infeasible and removing the `TODO`).
+Some of these, such as the handling of `TYP_SIMD8` in LSRA, may be addressed by other work items.
+
+### Add New Fixed-size Struct Types
+
+This would add `TYP_STRUCT1`, `TYP_STRUCT2`, `TYP_STRUCT4` and `TYP_STRUCT8` (the latter only on 64-bit
+systems). They would be normalized in the importer just like the SIMD types.
+  * First, simply add the types and ensure that they are handled as normal structs
+    * Do not make the transformation for structs containing a single ref
+      (see [Improve Struct Promotion](#Improve-Struct-Promotion)
+  * Next, fully enregister if there are no field accesses. This would be done in morph at the same time
+    that fully-enregistered SIMD types are marked as `lvRegStruct`.
+  * Next, modify struct promotion to fully promote (enregister) structs that have, by weight, 
+    more references as incoming or outgoing calls or returns, than field references.
+    * An initial investigation should be undertaken to determine if this is worthwhile.
+    * This requires codegen support to extract the fields from the register at their
+      references.
+
+###  <a name="Improve-Struct-Promotion"/>Improve Struct Promotion
+ * Aggressively promote lclVar struct incoming or outgoing args or returns whose fields match the ABI requirements.
+ * Aggressively promote pointer-sized fields of structs used as args or returns
+ * Allow struct promotion of locals that are passed or returned in a way that doesn't match
+   the field types.
+ * Fully promote structs of pointer size or less that have fewer field references than calls
+   or returns.
+ * Investigate whether it would be useful to re-type single-field structs, rather than creating new lclVars.
+   This would complicate type analysis when copied, passed or returned, but would avoid unnecessarily expanding
+   the lclVar data structures.
+
+### <a name="Block-Assignments"/>Improve and Simplify Block Assignment Morphing
+
+* `fgMorphOneAsgBlockOp` should probably be eliminated, and its functionality either moved to
```

Comment:
Tried. Main issue is that the current codegen for unrolled block ops is kind of poor. Codegen fix mostly done here: https://github.com/dotnet/coreclr/pull/21711